compiler/rustc_hir/src/hir.rs

```diff
@@ -2348,22 +2348,12 @@ pub enum UseKind {
 #[derive(Debug, HashStable_Generic)]
 pub struct TraitRef<'hir> {
     pub path: &'hir Path<'hir>,
+    pub def_id: Option<DefId>,
```

Comment:
There's no need to explicitly store the `DefId`. It's either a `Res::Def` of the appropriate kind or it's not. This should be an invariant of `TraitRef`, which you have done as part of HIR lowering, but should really be moved into a constructor for this type. At a later date, you could start treating unexpected `Res` variants as if they were `Res::Err`.